crates/ruma-api/src/lib.rs

```diff
@@ -591,3 +595,82 @@ impl MatrixVersion {
+pub enum EndpointPath {
+    /// Select an appropriate path version for a particular matrix version.
+    ///
+    /// Note: When using
+    /// [`try_into_http_request`](OutgoingRequest::try_into_http_request) with this variant, it can
+    /// emit additional errors;
+    /// - [`NoUnstablePath`](IntoHttpError::NoUnstablePath), if the matrix version is too new, and
+    ///   the endpoint has no defined `unstable` path (for maintenance or legacy reasons.)
+    /// - [`EndpointRemoved`](IntoHttpError::EndpointRemoved), if the matrix version is too old,
+    ///   and the endpoint is known to be removed in the same or a newer version.
+    ForMatrixVersion(MatrixVersion),
+
+    /// Select any path, preferring stable.
+    ///
+    /// Note: This does not emit any errors like
+    /// [`ForMatrixVersion`](EndpointPath::ForMatrixVersion) does, if the server's lowest matrix
+    /// version has removed the endpoint, or it has not yet stabilized the endpoint, this may
+    /// result in runtime HTTP errors, it is highly recommended to fetch `/versions`, extract
+    /// Ruma-known matrix versions, and pass the preferred version through `ForMatrixVersion`.
+    PreferStable,
+}
+
+impl From<MatrixVersion> for EndpointPath {
+    fn from(ver: MatrixVersion) -> Self {
+        Self::ForMatrixVersion(ver)
+    }
+}
+
+// This function needs to be public, yet hidden, as all `try_into_http_request`s would be using it.
+// We're also using MatrixVersion::repr() here, which is crate-private.
+#[doc(hidden)]
+pub fn select_path<'a>(
+    path: EndpointPath,
+    metadata: &'_ Metadata,
+    unstable: Option<fmt::Arguments<'a>>,
+    r0: Option<fmt::Arguments<'a>>,
+    stable: Option<fmt::Arguments<'a>>,
+) -> Result<fmt::Arguments<'a>, IntoHttpError> {
+    let version = match path {
+        EndpointPath::ForMatrixVersion(v) => v,
+        EndpointPath::PreferStable => {
+            return Ok(stable.or(r0).or(unstable).expect("one of three paths to be defined"))
+        }
+    };
+
+    if let Some(removed_ver) = metadata.removed {
+        if version.repr() >= removed_ver.repr() {
+            return Err(IntoHttpError::EndpointRemoved(version, removed_ver));
+        }
+    }
+
+    if let Some(depr_ver) = metadata.deprecated {
+        if version.repr() >= depr_ver.repr() {
+            // todo: emit warning
+        }
+    }
+
+    if let Some(added_ver) = metadata.added {
+        if version.repr() >= added_ver.repr() {
+            // r0 paths are "added" in V1_0, but if this has one, and the version is V1_0, then we
+            // use this fallback path
+            if let Some(r0) = r0 {
+                if version == MatrixVersion::V1_0 {
+                    return Ok(r0);
+                }
+            }
+
+            return Ok(stable.or(r0).expect(
+                "metadata.added is defined, so stable_path or r0_path must also be defined",
+            ));
+        }
+    }
+
+    unstable.ok_or(IntoHttpError::NoUnstablePath)
```

Comment:
This is quite a lot of nesting. There's probably a nicer way to do this, but I think it makes more sense to think about that once the question of fallibility of this function is resolved.